harness/determined/common/storage/boto3_credential_manager.py

```diff
@@ -0,0 +1,93 @@
+import logging
+from datetime import timedelta
+from typing import Any, Dict, Optional
+
+import boto3
+from boto3 import Session
+from botocore.credentials import (
+    CredentialProvider,
+    Credentials,
+    RefreshableCredentials,
+    SharedCredentialProvider,
+    _local_now,
+)
+from botocore.session import get_session
+
+
+class Boto3Manager:
+    """
+    Wrapper class around boto3 that allows registering of custom credential providers
+    and provides methods for setting the default boto3 session or fetching a new session
+    with the new context
+    """
+
+    def __init__(self) -> None:
+        self.session = get_session()
+        self._register_credential_provider(provider_name=SharedCredentialProvider.METHOD)
+
+    def _register_credential_provider(self, provider_name: str) -> None:
+        credential_chain = self.session.get_component("credential_provider")
+        credential_provider = credential_chain.get_provider(provider_name)
+        managed_credential_provider = RefreshedCredentialProvider(
+            credential_provider=credential_provider
+        )
+        credential_chain.insert_before(
+            name=provider_name, credential_provider=managed_credential_provider
+        )
+
+    def get_session(self) -> Session:
+        return boto3.Session(botocore_session=self.session)
+
+    def set_default(self) -> None:
+        boto3.setup_default_session(botocore_session=self.session)
+
+
+class RefreshedCredentialProvider(CredentialProvider):  # type: ignore
+    """
+    Creates a refreshable credential provider class given an existing credential provider in
+    the boto3 credential chain.
+
+    """
+
+    METHOD = "managed-refresh-cred"
+
+    def __init__(self, credential_provider: CredentialProvider, refresh_after: int = 60) -> None:
+        super().__init__()
+        self.refresh_after = timedelta(minutes=refresh_after)
+        self.credential_provider = credential_provider
+
+    def load(self) -> Optional[RefreshableCredentials]:
+        credentials = self._credentials()
+        return credentials and RefreshableCredentials(
+            method=self.METHOD,
+            refresh_using=self._refresh,
+            access_key=credentials.access_key,
+            secret_key=credentials.secret_key,
+            token=credentials.token,
+            expiry_time=self._expiry_time(),
+        )
+
+    def _refresh(self) -> Dict:
```

Comment:
I think what you've done here looks a lot like how RefreshableCredentials is used by the other parts of `botocore.credentials`.  But I think that it's not the right behavior for this situation.  I don't really think there should be a delay between when the files are rewritten and when we start using the new files.  The next time you try to read from the Credentials, I think the Credentials should refresh themselves from the file.  In that case you'd want to subclass Credentials with a different behavior from RefreshableCredentials.

Well... you could simulate the same behavior with a very short expiry time but where you use the timeout to check if the `SharedCredentialProvider.creds_filename` has been updated on disk, rather than always reloading every expiry time.  In practice that would probably be equivalent because you'd likely want a short backoff time with the first strategy anyway, rather than checking disk for every single request.  Ok, so I don't think you actually need to subclass Credentials.

But can we set the expiry time in such a way that we check for changes every minute or every 15 seconds or something?  Not reload the file every 15 seconds, but check if it's been modified on disk and reload if it has.  The user is presumably doing something like every 55 minutes, and if we have a timeout of 45 min, eventually a situation will arise where we check 5 minutes before they change it, which will leave us using an expired session for 40 minutes (assuming what they set expires 60 minutes after they set it, so old credentials are only valid for 5 minutes after they update to new ones).